src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -0,0 +1,553 @@
+            return modifiedRun;
+        }
+
+        private static bool UpdateCodeFlows(JArray codeFlows)
+        {
+            bool modifiedThreadFlowLocation = false;
+
+            foreach (JObject codeFlow in codeFlows)
+            {
+                JArray threadFlows = (JArray)codeFlow["threadFlows"];
+                if (threadFlows == null) { continue; }
+
+                foreach (JObject threadFlow in threadFlows)
+                {
+                    JArray threadFlowLocations = (JArray)threadFlow["locations"];
+                    if (threadFlowLocations == null) { continue; }
+
+                    foreach (JObject threadFlowLocation in threadFlowLocations)
+                    {
+                        // Update name of threadFlowLocation.executionTimeUtc: https://github.com/oasis-tcs/sarif-spec/issues/242
+                        modifiedThreadFlowLocation |= RenameProperty(threadFlowLocation, previousName: "timestamp", newName: "executionTimeUtc");
+
+                        // Delete threadFlowLocation.step property: https://github.com/oasis-tcs/sarif-spec/issues/203
+                        JToken step = threadFlowLocation["step"];
+                        if (step?.Type == JTokenType.Integer)
+                        {
+                            threadFlowLocation.Remove("step");
+                            modifiedThreadFlowLocation = true;
+                        }
+                    }
+                }
+            }
+            return modifiedThreadFlowLocation;
+        }
+
+        private static bool UpdateRunInvocations(JArray runs)
+        {
+            bool modifiedRun = false;
+
+            foreach (JObject run in runs)
+            {
+                JArray invocations = (JArray)run["invocations"];
+
+                if (invocations == null) { continue; }
+
+                foreach (JObject invocation in invocations)
+                {
+                    modifiedRun |= UpdateInvocationObject(invocation);
+                }
+            }
+            return modifiedRun;
+        }
+
+        internal static bool UpdateInvocationObject(JObject invocation)
+        {
+            bool modifiedInvocation = false;
+
+            // https://github.com/oasis-tcs/sarif-spec/issues/222
+            //
+            // Previously:
+            //     "workingDirectory": "/home/buildAgent/src",
+            // Now:
+            //     "workingDirectory": { "uri" : "/home/buildAgent/src" },
+
+            if (invocation["workingDirectory"]?.Type == JTokenType.String)
+            {
+                string workingDirectory = (string)invocation["workingDirectory"];
+
+                invocation.Property("workingDirectory").Remove();
+
+                var fileLocation = new JProperty("uri", workingDirectory);
+                var fileLocationObject = new JObject(fileLocation);
+                invocation["workingDirectory"] = fileLocationObject;
+
+                modifiedInvocation = true;
+            }
```

Comment:
Bonus points: DRY out by extracting a method `ReplaceUriStringWithFileLocation(JObject containing object, string propertyName)`. You have the same logic, expressed slightly differently in `UpdateOriginalUriBaseIds` below.

Reply:
There's actually very little shared code here. I did pull out the code in one place but the other didn't naturally utilize it. Oh well.

---
In reply to: [226765536](https://github.com/Microsoft/sarif-sdk/pull/1114#discussion_r226765536) [](ancestors = 226765536)